morf-integration-test/src/test/java/org/alfasoftware/morf/upgrade/TestDatabaseUpgradePathValidationService.java

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * Test for {@link DatabaseUpgradePathValidationService}
+ *
+ * @author Copyright (c) Alfa Financial Software Limited. 2023
+ */
+public class TestDatabaseUpgradePathValidationService {
+
+  @Rule
+  public MethodRule injectorRule = new InjectMembersRule(new TestingDataSourceModule());
+
+  @Inject
+  private SqlScriptExecutorProvider sqlScriptExecutorProvider;
+  @Inject
+  private DatabaseSchemaManager schemaManager;
+  @Inject
+  private DatabaseDataSetConsumer databaseDataSetConsumer;
+  @Inject
+  private ConnectionResources connectionResources;
+
+  @Inject
+  private DatabaseUpgradePathValidationService databaseUpgradePathValidationService;
+
+  private final Schema schema = SchemaUtils.schema(upgradeAuditTable());
+
+  private final DataSetProducer dataSet = dataSetProducer(schema)
+      .table(upgradeAuditTable().getName(),
+          record()
+              .setString("upgradeUUID", "003a64f7-3f9c-4624-8162-31606b9dad8b")
+              .setString("description", "Dummy upgrade 1")
+              .setLong("appliedTime", 20231113125538L),
+          record()
+              .setString("upgradeUUID", "ead354fd-a987-48dd-bd42-e1cebf9a3905")
+              .setString("description", "Dummy upgrade 2")
+              .setLong("appliedTime", 20231113125724L)
+      );
+
+
+  @Before
+  public void setup() {
+    schemaManager.dropAllTables();
+    schemaManager.mutateToSupportSchema(schema, DatabaseSchemaManager.TruncationBehavior.ALWAYS);
+    new DataSetConnector(dataSet, databaseDataSetConsumer).connect();
+  }
+
+
+  @After
+  public void tearDown() {
+    schemaManager.invalidateCache();
+  }
+
+
+  /**
+   * Test that the initialisation SQL produced provides an optimistic locking mechanism which prevents duplicate
+   * executions of the same script.
+   */
+  @Test
+  public void testInitialisationSql() {
+    // Given
+    List<String> initialisationSql = databaseUpgradePathValidationService.getPathValidationSql(2); // 2 records added to UpgradeAudit in test setup
+
+    // When
+    // Execute initialisation SQL for the first time (perform an upgrade)
+    sqlScriptExecutorProvider.get().execute(initialisationSql);
+
+    // Add a record to UpgradeAudit and drop the zzzUpgradeStatus table after first run as this is what would happen in an actual upgrade
+    addUpgradeAuditRecord();
+    dropUpgradeStatusTable();
+
+    // Then
+    RuntimeSqlException exception = assertThrows(RuntimeSqlException.class, () -> sqlScriptExecutorProvider.get().execute(initialisationSql));
+    // SQL State 23505 indicates a unique constraint violation on H2/PGSQL and 23000 on Oracle, so match based on either
+    String upgradeStatusUniqueConstraintViolationRegex = "Error executing SQL \\[INSERT INTO [A-Za-z0-9]*[_.]?zzzUpgradeStatus .*? SQL state \\[(23505|23000)]";
+    assertTrue("Should have been a unique constraint violation exception thrown", exception.getMessage().matches(upgradeStatusUniqueConstraintViolationRegex));
+  }

```

Comment:
That `SQL state \\[(23505|23000)]` might get different results on different platforms.
We have tested on H2/Postgres/Oracle, which is what we want at the moment.
This might need updating for other platforms in the future, if we start supporting them.

Note: There is a benefit to listing numbers here, we want to check for _unique constraint violation_ explicitly.
The query could fail for other reasons, for example, the table not being present, or disk being full, etc.